jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java

```diff
@@ -223,7 +224,13 @@ static String getPathInContext(Request request)
      */
     static long getTimeStamp(Request request)
     {
-        return System.currentTimeMillis() - NanoTime.millisSince(request.getHeadersNanoTime());
+        Long timeStampAttribute = (Long)request.getAttribute(TIMESTAMP_ATTRIBUTE);
+        if (timeStampAttribute != null)
+            return timeStampAttribute;
+
+        long timeStamp = System.currentTimeMillis() - NanoTime.millisSince(request.getHeadersNanoTime());
+        request.setAttribute(TIMESTAMP_ATTRIBUTE, timeStamp);
+        return timeStamp;
```

Comment:
I don't think we want an attribute for this.  Just create a field.

Actually it is not even worth the cost of a field to deflake a test.  Just make the test not check for an exact timestamp.  Perhaps parse the time and check it is within 1s

Reply:
> Perhaps parse the time and check it is within 1s

It could be more than one second out, its just more unlikely. And then it could just happen to be on the switch of the minute/hour/day so the less specific you make the test the less flaky it is, but there is still a chance it could fail.


Right now the `Request.getTimeStamp` return value can be completely wrong. So you can't really rely on the value. If we are to ignore this in the test we should probably add something to the javadoc to explain this.

Reply:
Note that it is likely to only ever be called by logging, which is the last thing ever done to the request.  So any effort to remember the value is wasted effort that will be done to ZILLIONS of requests forever for no good other than to make it easier for this unit test to pass!

Reply:
It is also used by various auditing and metrics libraries built on Jetty.

Reply:
@gregw `Request.getTimeStamp` is implemented as such:
```java
System.currentTimeMillis() - millisElapsed(request.getHeadersNanoTime(), NanoTime.now())
```
so the returned timestamp can vary by 1ms because of the rounding done in currentTimeMillis. No big deal.

But if there is a long GC pause between the `System.currentTimeMillis()` call and the `NanoTime.now()` one, the returned timestamp will be off by the GC pause duration. This _may_ be a bigger deal as a system with frequent long GC pauses can log imprecise timestamps.

Reply:
There is little we can do if we want both a nano time for accurate latency and a currentTimeMillis for an accurate wall clock.  No matter what, there could be a GC pause between the two samples so one will not correlate well with the other.
So even if at the time the very first byte arrives we grab both timestamps, then the second one could be delayed by a GC.

However, perhaps that is the least evil approach.   If we think this method may frequently be called, sometimes more than once, then we should grab both timestamps at the same place and have getters for both on the request.  At least that would give a mostly monotonically incrementing wall clock timestamp... but clock adjustments will still cause it to have some jiggle.   It is not a perfect measure no matter how you look at it.

So perhaps an extra field might be justified.... but an attribute is definitely not.